More Powerful Inspector

BelalNayzak opened the pull request:
🆕 What's Added
This pull request introduces a set of enhancements that improve both the usability and visual experience of the application:

----------------------------------

🌗 Dark/Light Mode Toggle
Automatically adapts to system theme or user preference.

📋 Click to Copy Content
Makes it easier to copy values such as headers, URLs, and JSON content with a single tap.

📂 Expandable/Collapsible Sections
Improves navigation by letting users toggle visibility of large sections like headers, request/response bodies, etc.

🌳 JSON Tree View
Added JSON-Tree display option with a structured, interactive tree view — expandable and more readable.

🎨 UI/UX Refinements
Added spacing, border radius, section highlights, and consistent padding for better layout and readability in both light and dark modes.

📌 Why These Changes
These features are aimed at making the developer experience more efficient and pleasant when inspecting requests and responses.

-------------------------------

Let me know if you have any comments. Happy to iterate based on feedback! 🙌

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -117,22 +118,27 @@ class RequestsInspector extends StatelessWidget {
   Widget build(BuildContext context) {
     return MaterialApp(
-      theme: ThemeData.dark().copyWith(
-        colorScheme: ColorScheme.dark(primary: Colors.grey[800]!),
-      ),
+      theme: context.watch<InspectorController>().isDarkMode
+          ? ThemeData.dark().copyWith(colorScheme: ColorScheme.dark(primary: Colors.grey[800]!))
+          : ThemeData.light(),
```
Better, wrap this part with Selector instead of converting the whole widget to StatefulWidget!

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
 
     return AppBar(
-      backgroundColor: Colors.black,
+      backgroundColor: context.read<InspectorController>().isDarkMode ? Colors.black : Colors.white,
```
And then we can use the selected value everywhere instead of making extra lookup using context just to get the `isDarkMode` value!

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
         onPressed: inspectorController.hideInspector,
         icon: const Icon(Icons.close),
-        color: Colors.white,
+        color: context.read<InspectorController>().isDarkMode ? Colors.white : Colors.black87,
```
same here

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
+            children: [
+              IconButton(
+                icon: context.read<InspectorController>().isTreeView
+                    ? Icon(Icons.account_tree_rounded, color: context.read<InspectorController>().isDarkMode ? Colors.white : Colors.black87, size: 20,)
+                    : Icon(Icons.account_tree_outlined, color: context.read<InspectorController>().isDarkMode ? Colors.white : Colors.black87, size: 20,),
```
same here

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
+                icon: context.read<InspectorController>().isTreeView
+                    ? Icon(Icons.account_tree_rounded, color: context.read<InspectorController>().isDarkMode ? Colors.white : Colors.black87, size: 20,)
+                    : Icon(Icons.account_tree_outlined, color: context.read<InspectorController>().isDarkMode ? Colors.white : Colors.black87, size: 20,),
+                onPressed: context.read<InspectorController>().toggleInspectorJsonView,
```
you can use `inspectorController ` directly instead!

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
+              ),
+
+              IconButton(
+                icon: context.read<InspectorController>().isDarkMode
```
same here

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
+          builder: (context, selectedTab, _) => Row(
+            children: [
+              IconButton(
+                icon: context.read<InspectorController>().isTreeView
```
same here

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -145,31 +151,58 @@ class _Inspector extends StatelessWidget {
+                icon: context.read<InspectorController>().isDarkMode
+                    ? const Icon(Icons.wb_sunny, color: Colors.white, size: 20,)
+                    : const Icon(Icons.brightness_2, color: Colors.black87, size: 20,),
+                onPressed: context.read<InspectorController>().toggleInspectorTheme,
```
You can use inspectorController directly here as well!

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -178,7 +211,7 @@ class _Inspector extends StatelessWidget {
   Widget _buildPopUpMenu(InspectorController inspectorController) {
     return PopupMenuButton(
-      icon: const Icon(Icons.more_vert, color: Colors.white),
+      icon: Icon(Icons.more_vert, color: context.read<InspectorController>().isDarkMode ? Colors.white : Colors.black87),
```
same here!

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -274,6 +309,7 @@ class _Inspector extends StatelessWidget {
     required bool isSelected,
     required bool isLeft,
     required VoidCallback onTap,
+    required BuildContext context,
```
Pass the values instead!

Abdelazeem777 replied:
isDark, primaryColor, and other...

Abdelazeem777 replied:
For better UI performance

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -282,22 +318,16 @@ class _Inspector extends StatelessWidget {
-                        ? const Radius.circular(0.0)
-                        : const Radius.circular(12.0),
-                  ),
+            color: context.read<InspectorController>().isDarkMode
+                ? isSelected ? Theme.of(context).primaryColor : Colors.black87
+                : isSelected ? Theme.of(context).primaryColor : Colors.white,
           ),
           child: Text(
```
Same here!

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -663,60 +781,71 @@ class _RequestDetailsPage extends StatelessWidget {
       text += '\nReceived at: $receivedTimeText\nDuration: $durationText';
     }
 
+    text += '\n\nURL: $url';
+
     return Padding(
       padding: const EdgeInsets.all(6.0),
-      child: Text(
+      child: SelectableText(
         text,
         style: const TextStyle(fontSize: 16.0),
       ),
     );
   }
 
-  Iterable<Widget> _buildHeadersBlock(headers) {
+  List<Widget> _buildHeadersBlock(BuildContext context, headers) {
```
Keep it Iterable, and no needs for BuildContext as the InspectorController is already Singleton

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -663,60 +781,71 @@ class _RequestDetailsPage extends StatelessWidget {
       text += '\nReceived at: $receivedTimeText\nDuration: $durationText';
     }
 
+    text += '\n\nURL: $url';
+
     return Padding(
       padding: const EdgeInsets.all(6.0),
-      child: Text(
+      child: SelectableText(
         text,
         style: const TextStyle(fontSize: 16.0),
       ),
     );
   }
 
-  Iterable<Widget> _buildHeadersBlock(headers) {
+  List<Widget> _buildHeadersBlock(BuildContext context, headers) {
     if (headers == null) return [];
     if ((headers is Map || headers is String || headers is List) &&
         headers.isEmpty) return [];
 
-    return [
-      _buildTitle('Headers:'),
-      _buildSelectableText(headers),
-    ];
+    if (context.read<InspectorController>().isTreeView) {
+      return [JsonTreeView(headers)];
+    }
+    else {
+      return [_buildSelectableText(headers)];
+    }
   }
 
-  Iterable<Widget> _buildQueryBlock(queryParameters) {
+  List<Widget> _buildQueryBlock(BuildContext context, queryParameters) {
```
Same

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -663,60 +781,71 @@ class _RequestDetailsPage extends StatelessWidget {
-  Iterable<Widget> _buildHeadersBlock(headers) {
+  List<Widget> _buildHeadersBlock(BuildContext context, headers) {
     if (headers == null) return [];
     if ((headers is Map || headers is String || headers is List) &&
         headers.isEmpty) return [];
 
-    return [
-      _buildTitle('Headers:'),
-      _buildSelectableText(headers),
-    ];
+    if (context.read<InspectorController>().isTreeView) {
+      return [JsonTreeView(headers)];
+    }
+    else {
+      return [_buildSelectableText(headers)];
+    }
   }
 
-  Iterable<Widget> _buildQueryBlock(queryParameters) {
+  List<Widget> _buildQueryBlock(BuildContext context, queryParameters) {
     if (queryParameters == null) return [];
     if ((queryParameters is Map ||
             queryParameters is String ||
             queryParameters is List) &&
         queryParameters.isEmpty) return [];
 
-    return [
-      _buildTitle('Parameters:'),
-      _buildSelectableText(queryParameters),
-    ];
+    if (context.read<InspectorController>().isTreeView) {
+      return [JsonTreeView(queryParameters)];
+    }
+    else {
+      return [_buildSelectableText(queryParameters)];
+    }
   }
 
-  Iterable<Widget> _buildRequestBodyBlock(requestBody) {
+  List<Widget> _buildRequestBodyBlock(BuildContext context, requestBody) {
```
Same

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -663,60 +781,71 @@ class _RequestDetailsPage extends StatelessWidget {
-  Iterable<Widget> _buildQueryBlock(queryParameters) {
+  List<Widget> _buildQueryBlock(BuildContext context, queryParameters) {
     if (queryParameters == null) return [];
     if ((queryParameters is Map ||
             queryParameters is String ||
             queryParameters is List) &&
         queryParameters.isEmpty) return [];
 
-    return [
-      _buildTitle('Parameters:'),
-      _buildSelectableText(queryParameters),
-    ];
+    if (context.read<InspectorController>().isTreeView) {
+      return [JsonTreeView(queryParameters)];
+    }
+    else {
+      return [_buildSelectableText(queryParameters)];
+    }
   }
 
-  Iterable<Widget> _buildRequestBodyBlock(requestBody) {
+  List<Widget> _buildRequestBodyBlock(BuildContext context, requestBody) {
     if (requestBody == null) return [];
     if ((requestBody is Map || requestBody is String || requestBody is List) &&
         requestBody.isEmpty) return [];
 
-    return [
-      _buildTitle('RequestBody:'),
-      _buildSelectableText(requestBody),
-    ];
+    if (context.read<InspectorController>().isTreeView) {
+      return [JsonTreeView(requestBody)];
+    }
+    else {
+      return [_buildSelectableText(requestBody)];
+    }
   }
 
-  Iterable<Widget> _buildResponseBodyBlock(responseBody) {
+  List<Widget> _buildResponseBodyBlock(BuildContext context, responseBody) {
```
Same

Abdelazeem777 commented:
<img width="378" alt="Screenshot 2025-05-02 at 2 17 57 PM" src="https://github.com/user-attachments/assets/56334188-9e54-4095-a51f-9091c9e66713" />
<img width="63" alt="Screenshot 2025-05-02 at 2 20 57 PM" src="https://github.com/user-attachments/assets/f7b6c5f8-74fd-4820-85e5-1575638111ad" />
<img width="63" alt="Screenshot 2025-05-02 at 2 21 06 PM" src="https://github.com/user-attachments/assets/d347e690-c4e0-44fa-9f40-db4255b44428" />

1. The Border not the selected request is still white in light mode
2. I think we need to make the default mode be the dark and pass the start mode from the constructor as optional but the default value is dark


<img width="377" alt="Screenshot 2025-05-02 at 2 23 42 PM" src="https://github.com/user-attachments/assets/9016ba78-80f5-4c02-8a13-bca463f1b05f" />

3. We need to move the the jsonview and dark mode buttons to the popup menu instead, becuase the "Requests Inspector" is not completely show


<img width="173" alt="Screenshot 2025-05-02 at 2 27 41 PM" src="https://github.com/user-attachments/assets/1a456f1c-bed1-447b-a467-5db1d43974ae" />

4. Too mush space between the expand button and copy


<img width="195" alt="Screenshot 2025-05-02 at 2 28 52 PM" src="https://github.com/user-attachments/assets/330d03cd-35af-4c31-a1c6-dad8bc747e48" />

5. Too much padding from the beginning


<img width="107" alt="Screenshot 2025-05-02 at 2 29 42 PM" src="https://github.com/user-attachments/assets/ff172bfa-d8c7-49df-ac5d-572ddc761731" />

6. The Border at edges is clipped


<img width="243" alt="Screenshot 2025-05-02 at 2 30 36 PM" src="https://github.com/user-attachments/assets/155e5a83-d428-4f2f-ad3a-4b13df9bd37e" />
<img width="431" alt="Screenshot 2025-05-02 at 2 31 08 PM" src="https://github.com/user-attachments/assets/d293e561-cef9-4c16-b38f-1e3cd231ff3e" />


7. I think we need the json view to be expanded by default at the beginning like what postman do

Abdelazeem777 commented:
We need to check the inspector view is still working fine with websockets and GraphQL responses

So please double check them and send the screenshots for the UI with them

BelalNayzak commented:
Hello @Abdelazeem777  👋,
How is your day going ✨

Thanks for your detailed feedback — I’ve iterated on all the points you mentioned and pushed the latest changes. 
I have also invested some time adding more enhancements and features in addition to your mentioned points.
Here's a summary of what’s been updated:

-------------------------------------------------------------------------------------

**🔧 Requested Changes Implemented**
- Applied all UI suggestions you provided in your last review.
- Applied all code suggestions you provided in your last review.
- Ensured consistency with existing codebase standards and code style.

**✨ Additional Improvements**
- Improved JsonTree experience with complex JSON.
- DataType color-sensitivity in JsonTree.
- Informative lists and maps JsonTree.
- Performance optimizations.
- Performed extra code refactoring and structuring for better readability and maintainability.
- Added several small UI enhancements to improve the UX.

**📄 README Updates**
- Documented the new features introduced.
- Added a Contributors section to acknowledge collaboration.

**🔀 Merged Changes**
- Integrated changes from your recent branch into this PR to avoid conflicts and ensure alignment.

-------------------------------------------------------------------------------------

Let me know if there’s anything else you’d like to tweak — happy to iterate further if needed!

Thanks!

Abdelazeem777 commented on lib/src/inspector_controller.dart:
```diff
@@ -208,4 +213,18 @@ class InspectorController extends ChangeNotifier {
     return _onStoppingResponse!(responseData);
   }
+
+  bool isDarkMode = true;
```
1. Better to be private and set getter to it
2. Move this var to the beginning of the class with the other vars

Abdelazeem777 commented on lib/src/json_tree_view_widget.dart:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  Widget _buildLeafNode(BuildContext context, String? key, dynamic value) {
+    return Consumer<InspectorController>(
```
Better using selector with bool on `isDarkMode`

Abdelazeem777 commented on lib/src/json_tree_view_widget.dart:
```diff
@@ -0,0 +1,332 @@
+
+  @override
+  Widget build(BuildContext context) {
+    return Consumer<InspectorController>(
+      builder: (context, controller, child) {
+        final isDarkMode = controller.isDarkMode;
```
Better using selector with bool on `isDarkMode`

Abdelazeem777 commented on lib/src/requests_inspector_widget.dart:
```diff
@@ -2,12 +2,30 @@ import 'dart:io';
+import 'package:requests_inspector/src/shared_widgets/inspector.dart';
 import '../requests_inspector.dart';
-import 'share_type_enum.dart';
+
+// Helper class for combining two values for a Selector
+// Placed at top-level for better organization
+class Tuple2<T1, T2> {
+  final T1 item1;
+  final T2 item2;
+
+  const Tuple2(this.item1, this.item2);
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is Tuple2 &&
+          runtimeType == other.runtimeType &&
+          item1 == other.item1 &&
+          item2 == other.item2;
+
+  @override
+  int get hashCode => item1.hashCode ^ item2.hashCode;
+}
```
No needs for this class, we can use dart built-in [records](https://dart.dev/language/records)

Abdelazeem777 commented on lib/src/shared_widgets/request_item.dart:
```diff
@@ -0,0 +1,82 @@
+  @override
+  Widget build(BuildContext context) {
+    // Crucial change: Use context.select to listen only to the selectedRequest for THIS item
+    final isSelected = context.select<InspectorController, bool>(
```
While we split it into new widget file, then better to pass the value from the outside!

Abdelazeem777 commented:
<img width="255" alt="Screenshot 2025-06-28 at 6 55 15 PM" src="https://github.com/user-attachments/assets/b7c390e7-409e-4ffe-93eb-158ed186534e" />
I don't like that purple color, tbh 😅
